rllib/models/specs/specs_dict.py

```diff
@@ -0,0 +1,66 @@
+from typing import Union, Type, Mapping, Any
+
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.models.specs.specs_base import TensorSpecs
+
+
+_MISSING_KEYS_FROM_SPEC = (
+    "The data dict does not match the model specs. Keys {} are "
+    "in the data dict but not on the given spec dict, and exact_match is set to True."
+)
+_MISSING_KEYS_FROM_DATA = (
+    "The data dict does not match the model specs. Keys {} are "
+    "in the spec dict but not on the data dict."
+)
+_TYPE_MISMATCH = (
+    "The data does not match the spec. The data element "
+    "{} has type {} (expected type {})."
+)
+
+SPEC_LEAF_TYPE = Union[Type, TensorSpecs]
+DATA_TYPE = Union[NestedDict[Any], Mapping[str, Any]]
+
+
+class ModelSpecDict(NestedDict[SPEC_LEAF_TYPE]):
+    """A NestedDict containing specs and class types."""
+
+    def validate(
+        self,
+        data: DATA_TYPE,
+        exact_match: bool = True,
+    ) -> None:
+        """Checks whether the data matches the spec.
+
+        Args:
+            data: The data which should match the spec. It can also be a spec.
+            exact_match: If true, the data and the spec must be exactly identical.
+                Otherwise, the data is validated as long as it contains at least the
+                elements of the spec, but can contain more entries.
+        Raises:
+            ValueError: If the data doesn't match the spec.
+        """
+        data = NestedDict(data)
+        missing_keys = set(self.keys()).difference(set(data.keys()))
+        if missing_keys:
+            raise ValueError(_MISSING_KEYS_FROM_DATA.format(missing_keys))
+        if exact_match:
+            data_spec_missing_keys = set(data.keys()).difference(set(self.keys()))
```

Comment:
buffer the keys in some local variable?
then kind of feel like missing_keys and data_spec_missing_keys are not necessary.

Reply:
even if I cache the keys to improve performance I still need A-B vs. B-A right?

Reply:
yes. A-B and B-A are fine.
I missed that you need the diffs for error messages.
please ignore, just cache the keys.